[Concurrency] Correct enqueue behavior in with... functions by adopting nonisolated(nonsending)

ktoso opened the pull request:
This is now subsumed by many other PRs, what remains to be done is the task group APIs.

----

Old description

The problem is that even with the #isolation parameter the non-Sendable
async closure operation _still_ would potentially hop off the caller
isolation. We introduced this change because the plan was the
non-Sendable closure would run on the isolated parameter's isolation,
but that's not actually the case: an "re-enqueue" can still happens in those cases.

Instead, we can use the nonisolated(nonsending) on the function and closure,
in order to guarantee there is no hop between those at all, and
developers can trust that adding this cancellation handler will not
cause any unexpected isolation changes and hops. 

The API was always documented to not hop as we execute the operation, 
so this brings the correct and expected behavior.

This corrects the following APIs:
- withCheckedContinuation
- withUnsafeContinuation
- withCheckedThrowingContinuation
- withUnsafeThrowingContinuation
- withTaskGroup
- withThrowingTaskGroup
- withDiscardingTaskGroup
- withThrowingDiscardingTaskGroup
- TaskLocal.withValue

as well as task group's next() and related ones...

This gets down the "re-enqueue" count of those methods by far, and makes all these more reliable and predictable.

resolves rdar://140110775
resolves rdar://162192512

ktoso commented:
This is blocked by incorrect behavior of the caller execution on a closure rdar://149107104

ktoso commented:
Now blocked on rdar://150017382

ktoso commented on stdlib/public/Concurrency/CMakeLists.txt:
```diff
@@ -63,6 +63,8 @@ endif()
   "-enable-experimental-feature"
   "IsolatedAny"
+  "-enable-experimental-feature"
+  "ExecutionAttribute"
```
we can remove this now

ktoso commented:
Good news, this was now unblocked by https://github.com/swiftlang/swift/pull/82858/

ktoso commented on stdlib/public/Concurrency/TaskCancellation.swift:
```diff
@@ -67,10 +67,26 @@ import Swift
-#if !$Embedded
-@backDeployed(before: SwiftStdlib 6.0)
-#endif
+@_alwaysEmitIntoClient
```
going with just AEIC those; we keep changing concurrency annotations and those methods are so small that AEIC for them is going to be fine anyway

ktoso commented on stdlib/public/Concurrency/TaskCancellation.swift:
```diff
@@ -67,10 +67,26 @@ import Swift
+  defer { unsafe _taskRemoveCancellationHandler(record: record) }
+
+  return try await operation()
+}
```
This way we stay on the caller for as long as possible, and the closure is also invoked on it. This improves the way we don't hop away from the caller unnecessarily anymore

ktoso commented:
This is almost unblocked, we need to fix throwing closures to not lose isolation -- rdar://155313349

ktoso commented:
@swift-ci please smoke test

ktoso commented:
@gottesmm will investigate the optimizer approach to solving this instead; seems we're missing the optimization in debug mode

ktoso commented:
@swift-ci please smoke test

jamieQ reviewed (commented):
some further unsolicited feedback/questions for your consideration :)

ktoso committed:
[Concurrency] Adopt noniso(nonsend) in many concurrency lib funcs

This drastically lessens the number of unexpected task enqueues and hops
that these APIs cause, and therefore improves performance and even
correctness in a few cases (like the withContinuation APIs)

ktoso committed:
use @export(implementation) instead of AEIC

ktoso committed:
avoid @export and use _alwaysEmitIntoClient instead

ktoso commented:
@swift-ci please smoke test

ktoso commented:
A bunch of real failures to follow up on:

17:00:49  Failed Tests (19):
17:00:49    Swift(macosx-x86_64) :: Concurrency/Reflection/reflect_task.swift
17:00:49    Swift(macosx-x86_64) :: Concurrency/Runtime/actor_counters.swift
17:00:49    Swift(macosx-x86_64) :: Concurrency/Runtime/actor_dynamic_subclass.swift
17:00:49    Swift(macosx-x86_64) :: Concurrency/Runtime/async_let_fibonacci.swift
17:00:49    Swift(macosx-x86_64) :: Concurrency/Runtime/async_task_executor_nsobject.swift
17:00:49    Swift(macosx-x86_64) :: Concurrency/Runtime/async_task_sleep.swift
17:00:49    Swift(macosx-x86_64) :: Concurrency/Runtime/async_task_sleep_cancel.swift
17:00:49    Swift(macosx-x86_64) :: Concurrency/Runtime/async_taskgroup_cancellation_race.swift
17:00:49    Swift(macosx-x86_64) :: Concurrency/Runtime/async_taskgroup_discarding_dontLeak.swift
17:00:49    Swift(macosx-x86_64) :: Concurrency/Runtime/async_taskgroup_discarding_neverConsumingTasks.swift
17:00:49    Swift(macosx-x86_64) :: Concurrency/Runtime/async_taskgroup_next_on_completed.swift
17:00:49    Swift(macosx-x86_64) :: Concurrency/Runtime/async_taskgroup_next_on_pending.swift
17:00:49    Swift(macosx-x86_64) :: Concurrency/Runtime/async_taskgroup_throw_rethrow.swift
17:00:49    Swift(macosx-x86_64) :: Concurrency/Runtime/cancellation_handler_concurrent.swift
17:00:49    Swift(macosx-x86_64) :: Concurrency/Runtime/sleep_executor.swift
17:00:49    Swift(macosx-x86_64) :: Concurrency/Runtime/startImmediately.swift
17:00:49    Swift(macosx-x86_64) :: Concurrency/Runtime/startImmediately_order.swift
17:00:49    Swift(macosx-x86_64) :: Concurrency/ShrinkWrap.swift
17:00:49    Swift(macosx-x86_64) :: Sanitizers/tsan/norace-task-group-cancellation.swift

Though at least we're past not compiling into runtime issues now, that's progress.

ktoso committed:
rebase fixes

ktoso commented:
More new compat issues uncovered

```

16:58:20  /home/build-user/swift-tools-protocols/Sources/SKLogging/LoggingScope.swift:92:71: error: cannot convert '@concurrent @Sendable () async throws -> Result' to 'nonisolated(nonsending) () async throws -> Result' because crossing of an isolation boundary requires parameter and result types to conform to 'Sendable' protocol
16:58:20   90 | ) async rethrows -> Result {
16:58:20   91 |   return try await LoggingScope.$_subsystem.withValue(subsystem) {
16:58:20   92 |     return try await LoggingScope.$_scope.withValue(scope, operation: operation)
16:58:20      |                                                                       |- error: cannot convert '@concurrent @Sendable () async throws -> Result' to 'nonisolated(nonsending) () async throws -> Result' because crossing of an isolation boundary requires parameter and result types to conform to 'Sendable' protocol
16:58:20      |                                                                       `- note: type 'Result' does not conform to 'Sendable' protocol
16:58:20   93 |   }
16:58:20   94 | }
```
